File: packages/ui-swiftui/Sources/KinetixUI/AppBar.swift
//
// AppBar.swift — KinetixAppBar (+ KinetixAppBarLink).
//
// Mirrors packages/ui/src/components/app-bar.tsx: a top application bar —
// a `brand` slot, a row of primary nav links (KinetixAppBarLink, `active`
// marks the current one) and a trailing `actions` slot, in a bordered
// header. The React component's md-breakpoint collapse-to-menu-toggle is
// dropped — a native top bar keeps the nav visible, horizontally
// scrollable if it overflows; same kind of scope-down as KinetixSheet
// (bottom-only). KinetixNavigationBar stays the mobile back-button bar.
//

import SwiftUI

public struct KinetixAppBar<Brand: View, Nav: View, Actions: View>: View {
    @Environment(\.kinetixColors) private var colors

    private let brand: Brand
    private let nav: Nav
    private let actions: Actions

    public init(
        @ViewBuilder brand: () -> Brand,
        @ViewBuilder nav: () -> Nav,
        @ViewBuilder actions: () -> Actions
    ) {
        self.brand = brand()
        self.nav = nav()
        self.actions = actions()
    }

    public var body: some View {
        HStack(spacing: 16) {
            brand
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 4) { nav }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            HStack(spacing: 8) { actions }
        }
        .padding(.horizontal, 16)
        .frame(height: 56) // h-14
        .frame(maxWidth: .infinity)
        .background(colors.background)
        .overlay(alignment: .bottom) {
            Rectangle().fill(colors.border).frame(height: 1) // border-b
        }
    }
}

public extension KinetixAppBar where Actions == EmptyView {
    init(@ViewBuilder brand: () -> Brand, @ViewBuilder nav: () -> Nav) {
        self.init(brand: brand, nav: nav, actions: { EmptyView() })
    }
}

/// A primary nav link inside a `KinetixAppBar`. `active` = the current destination.
public struct KinetixAppBarLink: View {
    @Environment(\.kinetixColors) private var colors

    private let label: String
    private let active: Bool
    private let action: () -> Void

    public init(_ label: String, active: Bool = false, action: @escaping () -> Void) {
        self.label = label
        self.active = active
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            Text(label)
                .font(.kinetixLabelMd)
                .foregroundStyle(active ? colors.foreground : colors.mutedForeground)
                .padding(.horizontal, 12) // px-3
                .padding(.vertical, 6) // py-1.5
                .background(active ? colors.accent : Color.clear)
                .clipShape(RoundedRectangle(cornerRadius: 6)) // rounded-md
        }
        .buttonStyle(.plain)
        .accessibilityAddTraits(active ? .isSelected : [])
    }
}

File: packages/ui-compose/ui/src/main/kotlin/com/kinetixui/ui/AppBar.kt
package com.kinetixui.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * KinetixAppBar family — mirrors `packages/ui/src/components/app-bar.tsx`:
 * a top application bar with a `brand` slot, a row of primary nav links
 * (`KinetixAppBarLink`, `active` marks the current one) and a trailing
 * `actions` slot, in a bordered header. The React component's
 * `md`-breakpoint collapse-to-menu-toggle is **dropped** — a native top
 * bar keeps the nav visible, horizontally scrollable if it overflows;
 * same kind of documented scope-down as `KinetixSheet` (bottom-only) or
 * `KinetixSidebar` (no rail mode). `KinetixNavigationBar` remains the
 * mobile back-button bar. Bottom border is hand-drawn with `drawBehind`,
 * same as `KinetixNavigationBar`/`KinetixFooter`. `h-14` (56dp) and the
 * link `px-3 py-1.5` (12dp / 6dp) aren't on the shared `spacing_*` scale
 * — hardcoded, same reasoning as `KinetixNavigationBar`'s off-scale sizes.
 */
@Composable
fun KinetixAppBar(
    modifier: Modifier = Modifier,
    brand: (@Composable () -> Unit)? = null,
    nav: (@Composable RowScope.() -> Unit)? = null,
    actions: (@Composable () -> Unit)? = null,
) {
    val colors = KinetixColorScheme.current
    val borderWidthPx = with(LocalDensity.current) {
        dimensionResource(R.dimen.border_width_default).toPx()
    }

    Row(
        modifier = modifier
            .fillMaxWidth()
            .height(56.dp) // h-14, not on the shared scale
            .background(colors.background)
            .drawBehind {
                drawLine(
                    color = colors.border,
                    start = Offset(0f, size.height),
                    end = Offset(size.width, size.height),
                    strokeWidth = borderWidthPx,
                )
            }
            .padding(horizontal = dimensionResource(R.dimen.spacing_4)),
        horizontalArrangement = Arrangement.spacedBy(dimensionResource(R.dimen.spacing_4)),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        if (brand != null) brand()
        if (nav != null) {
            Row(
                modifier = Modifier
                    .weight(1f)
                    .horizontalScroll(rememberScrollState()),
                horizontalArrangement = Arrangement.spacedBy(dimensionResource(R.dimen.spacing_1)),
                verticalAlignment = Alignment.CenterVertically,
                content = nav,
            )
        } else {
            Spacer(modifier = Modifier.weight(1f))
        }
        if (actions != null) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(dimensionResource(R.dimen.spacing_2)),
                verticalAlignment = Alignment.CenterVertically,
            ) { actions() }
        }
    }
}

/** A primary nav link inside a [KinetixAppBar]. `active` = the current destination. */
@Composable
fun KinetixAppBarLink(
    label: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    active: Boolean = false,
) {
    val colors = KinetixColorScheme.current
    Text(
        text = label,
        color = if (active) colors.foreground else colors.mutedForeground,
        fontWeight = FontWeight.Medium,
        fontSize = dimensionResource(R.dimen.font_size_body_sm).value.sp,
        modifier = modifier
            .clip(RoundedCornerShape(dimensionResource(R.dimen.radius_sm)))
            .background(if (active) colors.accent else Color.Transparent)
            .clickable(onClick = onClick)
            .padding(horizontal = 12.dp, vertical = 6.dp), // px-3 py-1.5, not on the shared scale
    )
}
